### ISSUE TITLE
Mapping Over Subworkflow Input Channel Twice Doubles Process Input Channel Size

### ISSUE THREAD

**Issue by jeffquinn-msk** [nextflow-io/nextflow#4444]
### Expected behavior and actual behavior

Given this workflow definition:

```
def create_chain_index_sequence(meta, n_chains) {
    output = []

    for (x in (0..(n_chains-1))) {
        output.add(tuple(meta, x))
    }
    return output
}

process MAKE_FILES {
    input:
    tuple val(meta), val(idx)


    output:
    tuple val(meta), path("*.out"), emit: fs

    script:
    """
    echo $idx > "${idx}.out"
    """
}

workflow TEST {
    take:
    ch_input

    main:
    process_input = ch_input.flatMap { create_chain_index_sequence(it[0], it[1]) }
    MAKE_FILES( process_input )

    process_input2 = ch_input.flatMap { create_chain_index_sequence(it[0], it[1]) }
    process_input2.view()
}


workflow {
   TEST( Channel.fromList([tuple([id:1], 5)]))
}
```

I would expect something like this output:

```
N E X T F L O W  ~  version 23.04.4
Launching `main.nf` [nauseous_maxwell] DSL2 - revision: b9694d5609
executor >  local (5)
[02/75a60f] process > TEST:MAKE_FILES (2) [100%] 5 of 5 ✔
[[id:1], 0]
[[id:1], 1]
[[id:1], 2]
[[id:1], 3]
[[id:1], 4]
```

Instead I get

```
N E X T F L O W  ~  version 23.04.4
Launching `main.nf` [nauseous_maxwell] DSL2 - revision: b9694d5609
executor >  local (10)
[02/75a60f] process > TEST:MAKE_FILES (2) [100%] 10 of 10 ✔
[[id:1], 0]
[[id:1], 0]
[[id:1], 1]
[[id:1], 1]
[[id:1], 2]
[[id:1], 2]
[[id:1], 3]
[[id:1], 4]
[[id:1], 3]
[[id:1], 4]
```

tl;dr: This behavior seems like surprising action at a distance: somehow the input channel to process, which was derived from mapping the input channel to my subworkflow, doubles in length if I map it again later on in my subworkflow.

### Steps to reproduce the problem

```
nextflow run main.nf
```

### Program output 

See above

### Environment 

* Nextflow version: version 23.04.4 build 5881
* Java version: OpenJDK 64-Bit Server VM (build 20+36-2344, mixed mode, sharing)
* Operating system: OSX 13.5 (22G74)
* Bash version: zsh 5.9 (x86_64-apple-darwin22.0)

### Additional context

None

**Comment by bentsherman** [comment]
It might be because in your function you declared `output` without a `def` qualifier, so it became a global variable. Try `def output = []` instead and see if it works.

See the warning at the end of [this section](https://nextflow.io/docs/latest/script.html#closures) in the docs
